staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -265,6 +266,13 @@ func (s *Serializer) unmarshal(into runtime.Object, data, originalData []byte) (
 		// to unmarshal directly into the object.
 		m := u.UnstructuredContent()
 		strictJSONErrs, err = kjson.UnmarshalStrict(data, &m)
+
+		// check for fatal errors first that won't get caught if unmarshaling
+		// directly into the object and are only detectable via a non-erroneous
+		// UnmarshalStrict call that results in no data being written to m
+		if err == nil && len(m) == 0 {
+			return nil, fmt.Errorf("unmarshal strict failed to unmarshal data: %v", data)
+		}
```

Comment:
this doesn't look right... data could be `{}`, right?

Reply:
That’s true. The issue we are trying to mitigate is that this check fails with a 500 error instead of a 400 error currently https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go#L356-L359 

When create is called on a the CRD client with an empty unstructured.Unstructured as the body, the server receives a request with a body “null\n” (sic).

The decoder then attempts to unmarshal this into an unstructured, which will fail _unless_ the unmarshal is strict, in which case it decodes directly into the body (which succeeds).

We then catch it later on, but throw a server-side missing kind error instead of a client-side invalid request error.

Another hacky solution is to unmarshal twice when strict unmarshaling first into the `into` to check for errors and second into the direct object to check for strict errs, but this doesn’t seem great either.

Reply:
Ok, I think I've got a much better solution here. We can just do a second unmarshal into `into` _after_ we know that nothing got written to `into`'s content, just to check for any errors.

I've added a test case for what previously failed as well as one for the case you've identified.